### PR TITLE
ci: update do_ci.sh targets and docs to the bazel test matrix.

### PR DESCRIPTION
**Pull request by htuch**
We've lost Mac OS X build information for do_ci.sh, since it's stale and I don't have a way to
verify it. Feel free to add in updated info later, but I think it might be better to use Bazel
directly here instead of via the container build.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,7 +0,0 @@
-DOCS_OUTPUT_DIR ?= generated/docs
```
Comment (mattklein123) [inline Makefile]: I'm not sure if anyone was relying on this to test docs locally. If we don't have it already, can we add a short section in the dev docs on how to build/view docs locally?

**Comment by mattklein123** [comment]
Per offline convo, ASAN build is much faster in bazel vs. cmake. Maybe it's due to parallelism? But also surprised we aren't destroying crappy travis VMs without setting --jobs.

**Comment by htuch** [comment]
It wasn't doing ASAN on `bazel.asan`, I think because of the fact I was building from a consumer project and missed the `bazel.rc`.

**Comment by htuch** [comment]
I'm going to move `asan` into `envoy_build_system.bzl`.

**Comment by mattklein123** [comment]
Ah OK, I had this same problem in our private build. I worked around by symlinking tools/bazel.rc -> envoy/tools/bazel.rc. +1 to building this into the bzl file.

**Comment by mattklein123** [comment]
This looks good, but I realized that we have various processing things that are done on the "normal" build like pushing images, etc. I think we need to fix all of that.

**Comment by mattklein123** [comment]
Also, do we care to understand why optimized build/test is now so slow? Seems like almost 2x regression from cmake.

**Comment by htuch** [comment]
I can confirm that locally (not on CI) that a `bazel test --jobs=2 //test/...` takes approximately 3x longer for `-c opt` than `fastbuild (14m 09 s vs. 5m 21s). I'll look at a `--profile` next locally to see what's going on.

**Comment by htuch** [comment]
Looking at individual C++ build commands, I can see the `-O2` build take 267% the time of the `-O0` build in one example (`test/mocks/router/mocks`). So, the build time that CI is reporting seems legit. What I wonder then is why it was so fast with cmake. Will look into what this was doing next.

**Comment by htuch** [comment]
`-c opt` builds both PIC and non-PIC for a reason I still can't figure out after reading https://groups.google.com/forum/#!topic/bazel-discuss/agycffdH0R0, hence doubling the test build time. Anyway, fixed by forcing the toolchain to be non-PIC in `cc_configure.bzl`.

**Comment by htuch** [comment]
Close/reopen due to a flake that might be in `async_client_impl_test`

File: ci/README.md [hunk 2]
```diff
@@ -13,71 +13,28 @@ binary built from the latest tip of master that passed tests.
...
+./ci/run_envoy_docker.sh bazel.fastbuild
 ```
 
-This bind mounts `<SOURCE_DIR>`, which allows for changes on the local
-filesystem to be consumed and outputs build artifacts in `<SOURCE_DIR>/build`.
-The static Envoy binary can be found in `<SOURCE_DIR>/build_debug/source/exe/envoy`.
+The output artifacts can be found in `/build/envoy/bazel-bin/source/exe/envoy-static` in the
+container.
 
 The `do_ci.sh` targets are:
```
Comment (mattklein123) [inline ci/README.md]: Should references to do_ci.sh in the docs be updated to point to run_envoy_docker.sh ?

**Comment by mattklein123** [comment]
@htuch you should be good to merge master.

**Comment by htuch** [comment]
@mattklein123 ready for final approval and merge.

**Pull request merged**